File: bbb/bbb/report/purchased_customer_list/purchased_customer_list.py
# Copyright (c) 2023, invento software limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns, data = get_columns(), get_customer_data(filters)
    return columns, data


def get_columns():
    """return columns"""
    columns = [
        {"label": _("Voucher No"), "fieldname": "name", "fieldtype": "Text", "options": "POS Invoice", "width": 170},
        {"label": _("Customer"), "fieldname": "customer_name", "fieldtype": "Link", "options": "Customer",
         "width": 100},
        {"label": _("Mobile No."), "fieldname": "customer_mobile_number", "fieldtype": "Text", "width": 150},
        {"label": _("Outlet"), "fieldname": "pos_profile", "fieldtype": "Text", "width": 120},
        {"label": _("Source"), "fieldname": "total_invoice", "fieldtype": "Int", "width": 80},
        {"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Text", "width": 550},
        {"label": _("Qty"), "fieldname": "qty", "fieldtype": "Int", "width": 60},
        {"label": _("  "), "fieldname": "click_here", "fieldtype": "Text", "width": 120},
    ]
    return columns


def get_conditions(filters):
    conditions = []

    if filters.get("from_date"):
        conditions.append("pos_invoice.posting_date <= '%s'" % filters.get("from_date"))

    if filters.get("to_date"):
        conditions.append("pos_invoice.posting_date >= '%s'" % filters.get("to_date"))

    if filters.get("company"):
        conditions.append("pos_invoice.company = '%s'" % filters.get("company"))

    if filters.get("pos_profile"):
        conditions.append("pos_invoice.pos_profile = '%s'" % filters.get("pos_profile"))

    conditions.append("pos_invoice.docstatus = 1")

    return " and ".join(conditions)


def get_customer_data(filters):
    conditions = get_conditions(filters)
    invoice_sql = """select pos_invoice.name, pos_invoice.customer_name, pos_invoice.customer_mobile_number, pos_invoice.pos_profile, pos_invoice.name as click_here from `tabPOS Invoice` pos_invoice where %s order by pos_invoice.name""" % conditions
    pos_invoices = frappe.db.sql(invoice_sql, as_dict=1)
    customer_data = []
    for invoice in pos_invoices:
        if filters.get('type') == 'Yesterday':
            feedback_status = frappe.db.get_value('Customer Feedback', {'voucher_no': invoice.get('name'), 'status': 'Customer Feedback Collected'}, 'name')
            if feedback_status:
                invoice['click_here'] = 'feedback_received'

        elif filters.get('type') == 'Follow-Up 1 Month':
            feedback_status = frappe.db.get_value('Customer Feedback', {'voucher_no': invoice.get('name'),
                                                                        'status': 'Completed'},
                                                  'name')
            if feedback_status:
                invoice['click_here'] = 'feedback_received'

        customer_data.append(invoice)
        item_sql = """select pos_invoice_item.item_name, pos_invoice_item.qty from `tabPOS Invoice Item` pos_invoice_item where pos_invoice_item.parent='%s' """ % invoice.get(
            'name')
        pos_invoice_items = frappe.db.sql(item_sql, as_dict=1)
        for pos_invoice_item in pos_invoice_items:
            customer_data.append(pos_invoice_item)
    return customer_data


@frappe.whitelist()
def create_new_doc(invoice_name):
    doc = frappe.new_doc("Customer Feedback")
    doc.voucher_no = invoice_name
    doc.insert()
    return doc.name

File: bbb/bbb/report/purchased_customer_list/purchased_customer_list.js
// Copyright (c) 2023, invento software limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Purchased Customer List"] = {
  "filters": [
    {
      "fieldname": "from_date",
      "label": __("From Date"),
      "fieldtype": "Date",
      "default": frappe.datetime.add_days(frappe.datetime.get_today(), -1),
      "reqd": 1,
    },
    {
      "fieldname": "to_date",
      "label": __("To Date"),
      "fieldtype": "Date",
      "default": frappe.datetime.add_days(frappe.datetime.get_today(), -1),
    },
    {
      "fieldname": "type",
      "label": __(""),
      "fieldtype": "Select",
      "default": "Yesterday",
      "options": ["Yesterday", "Follow-Up 1 Month"],
      "on_change": () => {
        let value = frappe.query_report.get_filter_value('type');
        if (value === "Yesterday") {
          frappe.query_report.set_filter_value('from_date', frappe.datetime.add_days(frappe.datetime.get_today(), -1));
          frappe.query_report.set_filter_value('to_date', frappe.datetime.add_days(frappe.datetime.get_today(), -1));
        } else if (value === "Follow-Up 1 Month") {
          frappe.query_report.set_filter_value('from_date', frappe.datetime.add_months(frappe.datetime.get_today(), -1));
          frappe.query_report.set_filter_value('to_date', frappe.datetime.add_months(frappe.datetime.get_today(), -1));
        }
        frappe.query_report.refresh()
      }
    },
    {
      "fieldname": "pos_profile",
      "label": __("Outlet"),
      "fieldtype": "Link",
      "options": "POS Profile"
    },
    {
      "fieldname": "company",
      "label": __("Company"),
      "fieldtype": "Link",
      "options": "Company",
      "default": frappe.defaults.get_default('company'),
      "width": "60px"
    },
  ],
  "initial_depth": 3,
  "tree": true,
  "parent_field": "voucher_no",
  "name_field": "child",
  "formatter": (value, row, column, data, default_formatter) => {
    value = default_formatter(value, row, column, data);
    if (column.fieldname === "customer_mobile_number") {
      value = `<div class="disabled" style="cursor: copy;" onclick="(()=> frappe.utils.copy_to_clipboard(${value}))()">${value.bold()}</div>`;
    }

    // console.log(frappe)
    if (column.fieldname === "click_here") {
      // value = `<a style='color:green' href='/app/new-customer-feedback' target='_blank' data-doctype='Customer Feedback'>${value.bold()}</a>`;
      if (value === 'feedback_received') {
        value = `<div style="  background-color: #f5f7fa; cursor: pointer; color: green" class="disabled"> ${value ? "Feedback Collected" : ""}</div>`;

      } else {
        value = `<div style="  background-color: #f5f7fa; cursor: pointer;" class="disabled" onclick="(()=> (frappe.query_reports['Purchased Customer List']).make_new_doc('` + value + `'))()"> ${value ? "Create Feedback" : ""}</div>`;

      }
    }
    return value;
  },
  "make_new_doc": (value) => {
    // window.open('/', '_blank')
    // setTimeout(function (){
    //   frappe.new_doc("Customer Feedback", {
    //     'voucher_no': value,
    //     'status' : 'Yesterday Purchased'
    //   },'_blank')
    // }, 1000)
    let report_type = frappe.query_report.get_filter_value('type');
    // if (report_type === 'Follow-Up 1 Month') {
    //   frappe.db.get_list('Customer Feedback', {
    //     fields: ['name'],
    //     filters: {
    //       voucher_no: value
    //     }
    //   }).then(records => {
    //     if (records) {
    //       // frappe.model.set_value('Customer Feedback', records[0]['name'], 'customer_feedback_section', 1)
    //       // frappe.model.set_value('Customer Feedback', records[0]['name'], 'product_feedback_section', 0)
    //       frappe.db.set_value('Customer Feedback', records[0]['name'], 'status', 'Follow-up Customer Feedback')
    //         .then(r => {
    //           let doc = r.message;
    //           var url = frappe.urllib.get_full_url('/desk/customer-feedback/' + doc.name);
    //           window.open(url, '_blank')
    //         })
    //     }
    //
    //   })
    // } else {
      frappe.db.insert({
        doctype: 'Customer Feedback',
        voucher_no: value,
        status: 'Purchased Customer Feedback',
      }).then(doc => {
        var url = frappe.urllib.get_full_url('/desk/customer-feedback/' + doc.name);
        window.open(url, '_blank')
      })
    // }
    // frappe.route_options = {
    //     'voucher_no': value,
    //     'status' : 'Yesterday Purchased'
    //   };
    // frappe.set_route('Form', 'Customer Feedback', 'new','_blank');
    // window.open('/app/customer-feedback/new-customer-feedback', '_blank')
  },
};
